File: Stack/minStack.java
class MinStack {

    Stack<Pair<Integer, Integer>> s;

    public MinStack() {
        s = new Stack<Pair<Integer, Integer>>();
    }

    public void push(int val) {
        if (s.isEmpty()) {
            Pair p = new Pair<>(val, val);
            s.push(p);
        } else {
            Pair p = new Pair<>(val, Math.min(val, this.getMin()));
            s.push(p);
        }
    }

    public void pop() {
        s.pop();
    }

    public int top() {
        return s.peek().getKey();
    }

    public int getMin() {
        return s.peek().getValue();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

// https://leetcode.com/problems/min-stack/

File: Stack/minStack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = []
        

    def push(self, val: int) -> None:
        self.stack.append(val)
        if (len(self.minStack) > 0):
            self.minStack.append(min(val, self.minStack[-1]))
        else:
            self.minStack.append(val)

    def pop(self) -> None:
        if (len(self.stack) > 0):
            self.minStack.pop()
            self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]

# https://leetcode.com/problems/min-stack